core/src/main/java/org/jruby/RubyArithmeticSequence.java

```diff
@@ -0,0 +1,565 @@
+        sequencec.defineAnnotatedMethods(RubyArithmeticSequence.class);
+
+        RubyClass seqMetaClass = sequencec.getMetaClass();
+        seqMetaClass.undefineMethod("new");
+
+        return sequencec;
+    }
+
+    public RubyArithmeticSequence(Ruby runtime, RubyClass klass) {
+        super(runtime, klass);
+    }
+
+    public RubyArithmeticSequence(Ruby runtime, RubyClass klass, IRubyObject generatedBy, String method, IRubyObject [] args, IRubyObject begin, IRubyObject end, IRubyObject step, IRubyObject excludeEnd) {
+        super(runtime, klass);
+        this.begin = begin;
+        this.end = end;
+        this.step = step;
+        this.excludeEnd = excludeEnd;
+        this.generatedBy = generatedBy;
+        this.method = method;
+        this.args = args;
+    }
+
+    public static RubyArithmeticSequence newArithmeticSequence(ThreadContext context, IRubyObject generatedBy, String method, IRubyObject [] args, IRubyObject begin, IRubyObject end, IRubyObject step, IRubyObject excludeEnd) {
+        return new RubyArithmeticSequence(context.runtime, context.runtime.getArithmeticSequence(), generatedBy, method, args, begin, end, step, excludeEnd);
+    }
+
+    // arith_seq_each
+    @JRubyMethod
+    public IRubyObject each(ThreadContext context, Block block) {
+        IRubyObject c = begin;
+        IRubyObject e = end;
+        IRubyObject s = step;
+        IRubyObject len_1, last;
+
+        if (!block.isGiven()) {
+            return this;
+        }
+
+        if (!(step instanceof RubyComplex) && floatStep(context, c, e, s, excludeEnd.isTrue(), true, block)) {
+            return this;
+        }
+
+        if (end.isNil()) {
+            while (true) {
+                block.yield(context, c);
+                c = ((RubyNumeric)c).op_plus(context, s);
+            }
+            //FIXME : unreachable statement
+            //return this;
+        }
+
+        if (Helpers.rbEqual(context, step, int2fix(context.runtime, 0)).isTrue()) {
+            while (true) {
+                block.yield(context, c);
+            }
+            //FIXME : unreachable statement
+            //return this;
+        }
+
+        len_1 = ((RubyNumeric)((RubyNumeric)e).op_minus(context, c)).idiv(context, s);
+        last = ((RubyNumeric)c).op_plus(context, Numeric.f_mul(context, s, len_1));
+        if (excludeEnd.isTrue() && Helpers.rbEqual(context, last, e).isTrue()) {
+            last = ((RubyNumeric)last).op_minus(context, s);
+        }
+
+        if (Numeric.f_negative_p(context, s)) {
+            while (RubyNumeric.numFuncall(context, c, new GeCallSite(), last).isTrue()) {
+                block.yield(context, c);
+                c = ((RubyNumeric)c).op_plus(context, s);
+            }
+        } else {
+            while (RubyNumeric.numFuncall(context, last, new GeCallSite(), c).isTrue()) {
+                block.yield(context, c);
+                c = ((RubyNumeric)c).op_plus(context, s);
+            }
+        }
```

Comment:
These call sites should be cached in JavaSites. This logic will construct a new one every time which will be slower than doing a slow path invocation.

This should ideally be done for *all* dynamic calls from Java logic. There's several other places in this PR that could benefit.

See following for examples of how to define and access those sites:

Definition: https://github.com/jruby/jruby/blob/master/core/src/main/java/org/jruby/runtime/JavaSites.java

Easy access via a static method: https://github.com/jruby/jruby/blob/master/core/src/main/java/org/jruby/RubyString.java#L6673

Usage: https://github.com/jruby/jruby/blob/master/core/src/main/java/org/jruby/RubyString.java#L4022-L4032